cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/profile/ProfileStore.java

```diff
@@ -82,94 +88,152 @@ public static void setupDatasets(DatasetFramework framework) throws IOException,
       MDSKey rowKey = getRowKey(profileId);
-      Profile value = getMDS(context).get(rowKey, Profile.class);
+      ProfileMeta value = mds.get(rowKey, ProfileMeta.class);
       if (value == null) {
         throw new NotFoundException(profileId);
       }
+      // TODO: https://issues.cask.co/browse/CDAP-13514, check for associations of the profile before deleting
+      if (value.getStatus() == ProfileStatus.ENABLED) {
+        throw new BadRequestException(String.format("Profile %s in namespace %s is currently enabled. A profile can " +
+                                                      "only be deleted if it is disabled",
+                                                    profileId.getProfile(), profileId.getNamespace()));
+      }
       mds.delete(getRowKey(profileId));
-    }, IOException.class, NotFoundException.class);
+    }, NotFoundException.class, BadRequestException.class);
   }
 
   /**
    * Delete all profiles in a given namespace.
    *
    * @param namespaceId the id of the namespace
-   * @throws IOException if there was an IO error deleting the profiles
    */
   public void deleteAllProfiles(NamespaceId namespaceId) throws IOException {
     Transactionals.execute(transactional, context -> {
       getMDS(context).deleteAll(getRowKey(namespaceId));
-    }, IOException.class);
+    });
+  }
+
+  public ProfileStatus getProfileStatus(ProfileId profileId) throws NotFoundException {
+    return getProfileInfo(profileId).getStatus();
+  }
+
+  /**
+   * Enable the profile. After the profile is enabled, any program/schedule can be associated with this profile.
+   *
+   * @param profileId the id of the profile to enable
+   * @throws NotFoundException if the profile is not found
+   * @throws ConflictException if the profile is already enabled
+   */
+  public void enableProfile(ProfileId profileId) throws NotFoundException, ConflictException {
+    changeProfileStatus(profileId, ProfileStatus.ENABLED);
+  }
+
+  /**
+   * Disable the profile. After the profile is disabled, any program/schedule cannot be associated with this profile.
+   *
+   * @param profileId the id of the profile to disable
+   * @throws NotFoundException if the profile is not found
+   * @throws ConflictException if the profile is already disabled
+   */
+  public void disableProfile(ProfileId profileId) throws NotFoundException, ConflictException {
+    changeProfileStatus(profileId, ProfileStatus.DISABLED);
+  }
+
+  private void changeProfileStatus(ProfileId profileId,
+                                   ProfileStatus expectedStatus) throws NotFoundException, ConflictException {
+    Transactionals.execute(transactional, context -> {
+      MetadataStoreDataset mds = getMDS(context);
+      MDSKey rowKey = getRowKey(profileId);
+      ProfileMeta oldMeta = mds.get(rowKey, ProfileMeta.class);
+      if (oldMeta == null) {
+        throw new NotFoundException(profileId);
+      }
+      if (oldMeta.getStatus() == expectedStatus) {
```

Comment:
Is there a reason that we want enabling an enabled profile to be an error? It seems more natural to me if we just return. Do you know what we do if you stop a stopped program?

Reply:
When we stop a stopped program, we will throw a 400 saying the program is not running. That's why I throw an exception here. Not sure what is the expected behavior though

Reply:
hm I see, then we should make this an error, though I still think 409 is more appropriate. A 400 is more like there is something wrong with the request, such that it will not succeed even on retry. 409 is more about the resource state being wrong, but a retry may succeed after fixing the state.

Can you open a jira about program start and stop being inconsistent? I think the right thing would be to change the stop call to return a 409 instead of a 400.

Reply:
Created JIRA: https://issues.cask.co/browse/CDAP-13530